pkg/readiness/object_tracker.go

```diff
@@ -126,7 +136,33 @@ func (t *objectTracker) CancelExpect(o runtime.Object) {
 	delete(t.expect, k)
 	delete(t.seen, k)
 	delete(t.satisfied, k)
-	t.cancelled[k] = struct{}{}
+	t.cancelled[k] = t.dataFactory()
+}
+
+func (t *objectTracker) TryCancelExpect(o runtime.Object) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	// Respect circuit-breaker.
+	if t.allSatisfied {
+		return
+	}
+
+	k, err := objKeyFromObject(o)
+	if err != nil {
+		log.Error(err, "skipping")
+		return
+	}
+
+	obj := t.expect[k]
+	shouldDelete := obj.decrementRetries()
+	t.expect[k] = obj
```

Comment:
Rather than copying the code from `CancelExpect`, I think we should call it directly.

We'll need to move `CancelExpect` to a different function that doesn't acquire a lock, say `cancelExpectNoLock` and rewrite `CancelExpect()` to acquire a lock and call the same function.

Reply:
Because of the `decrementRetries()` call in the middle, I had to pass that part in as a callback.  Let me know if you think it makes it too complicated.